testnet/stacks-node/src/event_dispatcher.rs

```diff
@@ -105,6 +123,40 @@ pub struct MinedNakamotoBlockEvent {
     pub tx_events: Vec<TransactionEvent>,
 }
 
+impl StackerDBChannel {
+    pub fn new() -> Self {
+        let (sender, recv_channel) = std::sync::mpsc::channel();
+        Self {
+            receiver: Mutex::new(Some(recv_channel)),
+            sender,
+        }
+    }
+
+    pub fn replace_receiver(&self, receiver: Receiver<StackerDBChunksEvent>) {
+        let mut guard = self
+            .receiver
+            .lock()
+            .expect("FATAL: poisoned StackerDBChannel lock");
+        guard.replace(receiver);
+    }
+
+    pub fn take_receiver(&self) -> Option<Receiver<StackerDBChunksEvent>> {
+        self.receiver
+            .lock()
+            .expect("FATAL: poisoned StackerDBChannel lock")
+            .take()
+    }
+
+    /// Is there a thread holding the receiver?
+    pub fn is_active(&self) -> bool {
+        // if the receiver field is empty (i.e., None), then a thread must have taken it.
+        self.receiver
+            .lock()
+            .expect("FATAL: poisoned StackerDBChannel lock")
+            .is_none()
```

Comment:
There's a race-condition in how this gets used, but the solution is pretty easy. Right now, the miner sends its block proposal to the StackerDB before taking the receiver channel. This means that if the signer set was doing its own coordination, they might start the protocol, sending messages to the miner coordinator before the miner coordinator took the channel, meaning those events would be lost the miner. Instead `take_receiver` should be invoked first, then the proposal broadcast. Right now, there's no actual race, though, because the miner coordinator is what initiates the signing protocol. But if the protocol changed such that signers individually sent back NACKs on reading the proposal before the miner submitted NonceRequests, those NACKs could end up being missed.

Reply:
If it's easy to fix, then can you just do it in this PR?

Reply:
But there's actually not a race for the reason mentioned.

Reply:
The "fix" would be sending the block proposal after instantiating the signer coordinator and obtaining the receiver channel -- this is easy enough to move around, but it would delay the message being sent slightly.